In [8]:
from pandasql import sqldf
import pandas as pd

# Create helper function for easier query execution
execute = lambda q: sqldf(q, globals())

# Load your CSV files into DataFrames
Summer_Medals = pd.read_csv("dataset/summer.csv")

# Execute query with a join and store the result
query = """
    SELECT *
    FROM Summer_Medals
"""
result_df = execute(query)

# Show results
result_df.head()

Year    City     Sport Discipline            Athlete Country Gender  \
0  1896  Athens  Aquatics   Swimming       HAJOS Alfred     HUN    Men   
1  1896  Athens  Aquatics   Swimming    HERSCHMANN Otto     AUT    Men   
2  1896  Athens  Aquatics   Swimming   DRIVAS Dimitrios     GRE    Men   
3  1896  Athens  Aquatics   Swimming  MALOKINIS Ioannis     GRE    Men   
4  1896  Athens  Aquatics   Swimming  CHASAPIS Spiridon     GRE    Men   

                        Event   Medal  
0              100M Freestyle    Gold  
1              100M Freestyle  Silver  
2  100M Freestyle For Sailors  Bronze  
3  100M Freestyle For Sailors    Gold  
4  100M Freestyle For Sailors  Silver

# Window functions vs GROUP BY

Which of the following is FALSE?


- Window functions can open a "window" to another table, whereas GROUP BY functions cannot.

# Numbering rows

The simplest application for window functions is numbering rows. Numbering rows allows you to easily fetch the nth row. For example, it would be very difficult to get the 35th row in any given table if you didn't have a column with each row's number.

In [11]:
query = """
SELECT
  *,
  -- Assign numbers to each row
  ROW_NUMBER() OVER() AS Row_N
FROM Summer_Medals
ORDER BY Row_N ASC;
"""
result_df = execute(query)

# Show results
result_df.head()

Year    City     Sport Discipline            Athlete Country Gender  \
0  1896  Athens  Aquatics   Swimming       HAJOS Alfred     HUN    Men   
1  1896  Athens  Aquatics   Swimming    HERSCHMANN Otto     AUT    Men   
2  1896  Athens  Aquatics   Swimming   DRIVAS Dimitrios     GRE    Men   
3  1896  Athens  Aquatics   Swimming  MALOKINIS Ioannis     GRE    Men   
4  1896  Athens  Aquatics   Swimming  CHASAPIS Spiridon     GRE    Men   

                        Event   Medal  Row_N  
0              100M Freestyle    Gold      1  
1              100M Freestyle  Silver      2  
2  100M Freestyle For Sailors  Bronze      3  
3  100M Freestyle For Sailors    Gold      4  
4  100M Freestyle For Sailors  Silver      5

# Numbering Olympic games in ascending order

The Summer Olympics dataset contains the results of the games between 1896 and 2012. The first Summer Olympics were held in 1896, the second in 1900, and so on. What if you want to easily query the table to see in which year the 13th Summer Olympics were held? You'd need to number the rows for that.

In [13]:
query = """
  SELECT DISTINCT year
  FROM Summer_Medals
  ORDER BY Year ASC
"""
result_df = execute(query)

# Show results
result_df.head()

Year
0  1896
1  1900
2  1904
3  1908
4  1912

In [12]:
query = """
SELECT
  Year,
  -- Assign numbers to each year
  ROW_NUMBER() OVER() AS Row_N
FROM (
  SELECT DISTINCT year
  FROM Summer_Medals
  ORDER BY Year ASC
) AS Years
ORDER BY Year ASC;

"""
result_df = execute(query)

# Show results
result_df.head()

year  Row_N
0  1896      1
1  1900      2
2  1904      3
3  1908      4
4  1912      5

# Numbering Olympic games in descending order

You've already numbered the rows in the Summer Medals dataset. What if you need to reverse the row numbers so that the most recent Olympic games' rows have a lower number?

In [14]:
query = """
SELECT
  Year,
  -- Assign the lowest numbers to the most recent years
  ROW_NUMBER() OVER (ORDER BY Year DESC) AS Row_N
FROM (
  SELECT DISTINCT Year
  FROM Summer_Medals
) AS Years
ORDER BY Year;

"""
result_df = execute(query)

# Show results
result_df.head()

Year  Row_N
0  1896     27
1  1900     26
2  1904     25
3  1908     24
4  1912     23

# Numbering Olympic athletes by medals earned

Row numbering can also be used for ranking. For example, numbering rows and ordering by the count of medals each athlete earned in the OVER clause will assign 1 to the highest-earning medalist, 2 to the second highest-earning medalist, and so on.

In [15]:
query = """
SELECT
  -- Count the number of medals each athlete has earned
  Athlete,
  COUNT(medal) AS Medals
FROM Summer_Medals
GROUP BY Athlete
ORDER BY Medals DESC;

"""
result_df = execute(query)

# Show results
result_df.head()

Athlete  Medals
0     PHELPS Michael      22
1    LATYNINA Larisa      18
2  ANDRIANOV Nikolay      15
3     SHAKHLIN Boris      13
4        ONO Takashi      13

In [16]:
query = """
WITH Athlete_Medals AS (
  SELECT
    -- Count the number of medals each athlete has earned
    Athlete,
    COUNT(*) AS Medals
  FROM Summer_Medals
  GROUP BY Athlete)

SELECT
  -- Number each athlete by how many medals they've earned
  Athlete,
  ROW_NUMBER() OVER (ORDER BY medals DESC) AS Row_N
FROM Athlete_Medals
ORDER BY Medals DESC;
"""
result_df = execute(query)

# Show results
result_df.head()

Athlete  Row_N
0       PHELPS Michael      1
1      LATYNINA Larisa      2
2    ANDRIANOV Nikolay      3
3  MANGIAROTTI Edoardo      4
4          ONO Takashi      5

# Reigning weightlifting champions

A reigning champion is a champion who's won both the previous and current years' competitions. To determine if a champion is reigning, the previous and current years' results need to be in the same row, in two different columns.

In [17]:
query = """
SELECT
  -- Return each year's champions' countries
  year,
  country AS champion
FROM Summer_Medals
WHERE
  Discipline = 'Weightlifting' AND
  Event = '69KG' AND
  Gender = 'Men' AND
  Medal = 'Gold';
"""
result_df = execute(query)

# Show results
result_df.head()

Year champion
0  2000      BUL
1  2004      CHN
2  2008      CHN
3  2012      CHN

In [18]:
query = """
WITH Weightlifting_Gold AS (
  SELECT
    -- Return each year's champions' countries
    Year,
    Country AS champion
  FROM Summer_Medals
  WHERE
    Discipline = 'Weightlifting' AND
    Event = '69KG' AND
    Gender = 'Men' AND
    Medal = 'Gold')

SELECT
  Year, Champion,
  -- Fetch the previous year's champion
  LAG(Champion) OVER
    (ORDER BY Year ASC) AS Last_Champion
FROM Weightlifting_Gold
ORDER BY Year ASC;
"""
result_df = execute(query)

# Show results
result_df.head()

Year champion Last_Champion
0  2000      BUL          None
1  2004      CHN           BUL
2  2008      CHN           CHN
3  2012      CHN           CHN

# Reigning champions by gender

You've already fetched the previous year's champion for one event. However, if you have multiple events, genders, or other metrics as columns, you'll need to split your table into partitions to avoid having a champion from one event or gender appear as the previous champion of another event or gender.

In [21]:
query = """
  SELECT DISTINCT
    Gender, Year, Country
  FROM Summer_Medals
  WHERE
    Year >= 2000 AND
    Event = 'Javelin Throw' AND
    Medal = 'Gold'
  ORDER BY Gender ASC, Year ASC;
"""
result_df = execute(query)

# Show results
result_df.head()

Gender  Year Country
0    Men  2000     CZE
1    Men  2004     NOR
2    Men  2008     NOR
3    Men  2012     TTO
4  Women  2000     NOR

In [19]:
query = """
WITH Tennis_Gold AS (
  SELECT DISTINCT
    Gender, Year, Country
  FROM Summer_Medals
  WHERE
    Year >= 2000 AND
    Event = 'Javelin Throw' AND
    Medal = 'Gold')

SELECT
  Gender, Year,
  Country AS Champion,
  -- Fetch the previous year's champion by gender
  LAG(Country) OVER (PARTITION BY gender
            ORDER BY year ASC) AS Last_Champion
FROM Tennis_Gold
ORDER BY Gender ASC, Year ASC;
"""
result_df = execute(query)

# Show results
result_df.head()

Gender  Year Champion Last_Champion
0    Men  2000      CZE          None
1    Men  2004      NOR           CZE
2    Men  2008      NOR           NOR
3    Men  2012      TTO           NOR
4  Women  2000      NOR          None

# Reigning champions by gender and event

In the previous exercise, you partitioned by gender to ensure that data about one gender doesn't get mixed into data about the other gender. If you have multiple columns, however, partitioning by only one of them will still mix the results of the other columns.

In [24]:
query = """
  SELECT DISTINCT
    Gender, Year, Event, Country
  FROM Summer_Medals
  WHERE
    Year >= 2000 AND
    Discipline = 'Athletics' AND
    Event IN ('100M', '10000M') AND
    Medal = 'Gold'
    ORDER BY Event ASC, Gender ASC, Year ASC;
"""
result_df = execute(query)

# Show results
result_df.head()

Gender  Year   Event Country
0    Men  2000  10000M     ETH
1    Men  2004  10000M     ETH
2    Men  2008  10000M     ETH
3    Men  2012  10000M     GBR
4  Women  2000  10000M     ETH

In [22]:
query = """
WITH Athletics_Gold AS (
  SELECT DISTINCT
    Gender, Year, Event, Country
  FROM Summer_Medals
  WHERE
    Year >= 2000 AND
    Discipline = 'Athletics' AND
    Event IN ('100M', '10000M') AND
    Medal = 'Gold')

SELECT
  Gender, Year, Event,
  Country AS Champion,
  -- Fetch the previous year's champion by gender and event
  LAG(Country) OVER (PARTITION BY gender, event
            ORDER BY Year ASC) AS Last_Champion
FROM Athletics_Gold
ORDER BY Event ASC, Gender ASC, Year ASC;
"""
result_df = execute(query)

# Show results
result_df.head()

Gender  Year   Event Champion Last_Champion
0    Men  2000  10000M      ETH          None
1    Men  2004  10000M      ETH           ETH
2    Men  2008  10000M      ETH           ETH
3    Men  2012  10000M      GBR           ETH
4  Women  2000  10000M      ETH          None

# Row numbers with partitioning

If you run `ROW_NUMBER() OVER (PARTITION BY Year ORDER BY Medals DESC)` on the following table, what row number would the 2008 Iranian record have?
```
| Year | Country | Medals |
|------|---------|--------|
| 2004 | IRN     | 32     |
| 2004 | LBN     | 17     |
| 2004 | KSA     | 4      |
| 2008 | IRQ     | 29     |
| 2008 | IRN     | 27     |
| 2008 | UAE     | 12     |
```

- 2